src/main/java/com/celements/appScript/AppScriptService.java

```diff
@@ -87,14 +131,42 @@ public DocumentReference getAppScriptDocRef(String scriptName) {
     }
   }
 
+  @Override
+  public DocumentReference getAppRecursiveScriptDocRef(String scriptName) {
+    String scriptNameFound = findAppScriptRecursivly(scriptName,
+        sn -> hasLocalAppRecursiveScript(sn)
+            || hasCentralAppRecursiveScript(sn),
+        sn -> sn.lastIndexOf("/") > 0);
+    if (hasLocalAppRecursiveScript(scriptNameFound)) {
+      return getLocalAppRecursiveScriptDocRef(scriptNameFound);
+    } else if (hasCentralAppRecursiveScript(scriptNameFound)) {
+      return getCentralAppRecursiveScriptDocRef(scriptNameFound);
+    }
+    return null;
+  }
+
   @Override
   public DocumentReference getLocalAppScriptDocRef(String scriptName) {
-    return new DocumentReference(getContext().getDatabase(), APP_SCRIPT_SPACE_NAME, scriptName);
+    return RefBuilder.from(mContext.getWikiRef()).space(APP_SCRIPT_SPACE_NAME).doc(scriptName)
+        .build(DocumentReference.class);
   }
 
   @Override
   public DocumentReference getCentralAppScriptDocRef(String scriptName) {
-    return new DocumentReference("celements2web", APP_SCRIPT_SPACE_NAME, scriptName);
+    return RefBuilder.from(XWikiConstant.CENTRAL_WIKI).space(APP_SCRIPT_SPACE_NAME).doc(scriptName)
+        .build(DocumentReference.class);
+  }
+
+  @Override
+  public DocumentReference getLocalAppRecursiveScriptDocRef(String scriptName) {
+    return RefBuilder.from(mContext.getWikiRef()).space(APP_RECURSIVE_SCRIPT_SPACE_NAME)
+        .doc(scriptName).build(DocumentReference.class);
+  }
+
+  @Override
+  public DocumentReference getCentralAppRecursiveScriptDocRef(String scriptName) {
+    return RefBuilder.from(XWikiConstant.CENTRAL_WIKI).space(APP_RECURSIVE_SCRIPT_SPACE_NAME)
+        .doc(scriptName).build(DocumentReference.class);
```

Comment:
just fyi these: these methods throw IAE when the scriptName is empty or null. if fast failure isn't desired consider using `buildOpt` instead to get an optional.

Reply:
done.